### PR TITLE
Implementation of Carlini&Wagner's L0 optimized attack

### LINKED ISSUE #443: Implement Carlini & Wagner L_0 Attack
_Carlini and Wagner_ introduced an interesting algorithm for the generation of targeted adversarial examples, their method defeats _Defensive distillation_ and other defense methods.
The authors define a new objective function that can be minimized with the distance (the paper introduces 7 objective function candidates). They also introduce a new variant _ω_ to avoid the box constraint. General optimizers like _Adam_ and _SGD_ were used to find adversarial examples.
The authors have adapted their method for the `L_0`, `L_2` and `L_inf` distance metrics. The `L_0` attack is complex since this distance metric is non-differentiable and therefore is ill-suited for standard gradient descent. Instead, the authors introduced an iterative method that calls the `L_2` attack to generate adversarial examples using a subset of attributes and then computes the derivatives to identify the attributes that contributed the least and fix them. 
ART provides an implementation Carlini & Wagner `L_2` and `L_inf ` attacks only. It would be interesting to provide an implementation for the `L_0` attacks.

Paper link : https://arxiv.org/abs/1608.04644
GitHub link : https://github.com/carlini/nn_robust_attacks/

### PR DESCRIPTION
**Pull request by mamerzouk**
# Description

Implementation of Carlini&Wagner's L0 optimized attack.

Fixes #443 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Generating adversarial examples on the intrusion detection dataset NSL-KDD.
- [x] Generating adversarial examples on the intrusion detection dataset CIDDS-001.

**Test Configuration**:
- Fedora 31 (Linux 5.3.7-301.fc31.
- Python version 3.7.
- ART dev_1.6.
- PyTorch version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @mamerzouk Thank you very much for this great PR! I have changed the target branch to `dev_1.6.0` to include the new attack as a new feature in the next minor release. We'll do a format review soon!

File: art/attacks/evasion/carlini.py [hunk 1]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+                clip_max,
+            )
+            perturbation_L1_norm = np.abs(x_adv - x)
+
+            # gradient * perturbation tells how much reduction to the objective function we obtain for each attribute
+            objective_reduction = np.abs(objective_loss_gradient) * perturbation_L1_norm
+
+            # Put a huge nomber as objective_reduction value for fixed feature (in order not to select them again)
+            # was np.inf before, but inf * 0 = nan
+            objective_reduction += 999999999999999 * (activation == 0).astype(int)
```
Comment (mathsinn) [inline art/attacks/evasion/carlini.py:1298]: I'd suggest to find a way for adding `np.inf` (e.g. using `np.where`) instead of resorting to hard-coded constant, if possible.

**Review by mathsinn — CHANGES_REQUESTED** [review]
Thank you @mamerzouk for this PR! I left some comments - could you please take a look?

Also, would you be interested in adding your affiliation to our AUTHORS file? https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/AUTHORS

File: art/attacks/evasion/carlini.py [hunk 2]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+
+
+class CarliniL0Method(EvasionAttack):
+    """
+    The L_0 distance metric is non-differentiable and therefore is ill-suited for standard gradient descent.
+    Instead, we use an iterative algorithm that, in each iteration, identifies some pixels that don’t have much effect
+    on the classifier output and then fixes those pixels, so their value will never be changed. The set of fixed pixels
+    grows in each iteration until we have, by process of elimination, identified a minimal (but possibly not minimum)
+    subset of pixels that can be modified to generate an adversarial example. In each iteration, we use our L_2 attack
+    to identify which pixels are unimportant.
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Thank you for your comments, I fixed it. I hope the citation format is appropriate.

File: art/attacks/evasion/carlini.py [hunk 3]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
 
         if not isinstance(self.batch_size, (int, np.int)) or self.batch_size < 1:
             raise ValueError("The batch size must be an integer greater than zero.")
+
+
+class CarliniL0Method(EvasionAttack):
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Oh good idea, thank you. I followed your suggestion.

File: art/attacks/evasion/carlini.py [hunk 4]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+
+    attack_params = EvasionAttack.attack_params + [
+        "confidence",
+        "targeted",
+        "learning_rate",
+        "max_iter",
+        "binary_search_steps",
+        "initial_const",
+        "mask",
+        "warm_start",
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py:822]: Thank you, I implemented the warm start to make it useful.

File: art/attacks/evasion/carlini.py [hunk 5]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+        :param initial_const: The initial trade-off constant `c` to use to tune the relative importance of distance and
+                confidence. If `binary_search_steps` is large, the initial constant is not important, as discussed in
+                Carlini and Wagner (2016).
+        :type initial_const: `float`
+        :param mask: The initial features that can be modified by the algorithm. If not specified, the
+                algorithm uses the full feature set.
+        :type mask: `np.ndarray`
+        :param warm_start: Instead of starting gradien descent in each iteration from the initial image. we start the
+                gradient descent from the solution found on the previous iteration.
+        :type warm_start: `boolean`
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py:874]: Thank you, it is implemented now.

File: art/attacks/evasion/carlini.py [hunk 6]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+        # It appears this is what Carlini and Wagner (2016) are alluding to in their footnote 8. However, it is not
+        # clear how their proposed trick ("instead of scaling by 1/2 we scale by 1/2 + eps") works in detail.
+        self._tanh_smoother = 0.999999
+
+        # The tanh transformation does not always map inputs back to their original values event if they're unmodified
+        # To overcom this problem, we set a threshold of minimal diffrence considered as perturbation
+        # Below this threshold, a diffrence between values is considered as tanh transformation diffrence
+        self._perturbation_threshold = 1e-06
+
+    def _loss(self, x, x_adv, target, c_weight):
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Thank you, it is implemented now.

File: art/attacks/evasion/carlini.py [hunk 7]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+                c_final,
+                clip_min,
+                clip_max,
+            )
+            perturbation_L1_norm = np.abs(x_adv - x)
+
+            # gradient * perturbation tells how much reduction to the objective function we obtain for each attribute
+            objective_reduction = np.abs(objective_loss_gradient) * perturbation_L1_norm
+
+            # Put a huge nomber as objective_reduction value for fixed feature (in order not to select them again)
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: When French and English get mixed up in my head. Thank you.

File: art/attacks/evasion/carlini.py [hunk 8]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+
+            logger.info(
+                "Success rate of C&W L_2 attack: %.2f%%",
+                100 * compute_success(self.estimator, x, y, x_adv, self.targeted, batch_size=self.batch_size),
+            )
+
+            # If the L_2 attack can't find any adversarial examples with the new activation, return the last one
+            z_logits, l2dist, loss = self._loss(x, x_adv, y, c_final)
+            attack_success = loss - l2dist <= 0
+            # l0dist = np.sum((np.abs(x_batch - x_adv_batch) > self._perturbation_threshold).astype(int), axis=1)
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Nice catch, thank you.

File: art/attacks/evasion/carlini.py [hunk 9]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+            y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+
+        if self.mask is None:
+            # No initial activation provided, use the full feature set
+            activation = np.ones(x.shape)
+        else:
+            # Check if the initial activation has the same dimension as the input data
+            if self.mask.shape != x.shape:
+                raise ValueError("The mask must have the same dimensions as the input data.")
+            activation = np.array(self.mask).astype(float)
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py:949]: I did not understand very well this comment, I thought the mask was used in this implementation.

File: art/attacks/evasion/carlini.py [hunk 10]
```diff
@@ -797,3 +797,563 @@ def _check_params(self) -> None:
...
+                # The optimization is performed in tanh space to keep the adversarial images bounded in correct range
+                x_batch_tanh = original_to_tanh(x_batch, clip_min, clip_max, self._tanh_smoother)
+
+                # Initialize binary search:
+                c_current = self.initial_const * np.ones(x_batch.shape[0])
+                c_lower_bound = np.zeros(x_batch.shape[0])
+                c_double = np.ones(x_batch.shape[0]) > 0
+
+                # Initialize placeholders for best l2 distance and attack found so far
+                best_l2dist = np.inf * np.ones(x_batch.shape[0])
```
[the remark this answers is not in the document]
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Sure! Thank you for the remark.

File: art/attacks/evasion/carlini.py (hunk 1 above)
Comment (mamerzouk) [reply art/attacks/evasion/carlini.py]: Good idea, `np.where()` does the job perfectly, thank you.

File: art/attacks/evasion/carlini.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mathsinn) [reply art/attacks/evasion/carlini.py:949]: You're right, now I see where it's coming from.

File: art/attacks/evasion/carlini.py [hunk 11]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
 
         if not isinstance(self.batch_size, (int, np.int)) or self.batch_size < 1:
             raise ValueError("The batch size must be an integer greater than zero.")
+
+
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:801]:
```suggestion

# pylint: disable=C0302
```

File: art/attacks/evasion/carlini.py [hunk 12]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
...
+        :type initial_const: `float`
+        :param max_halving: Maximum number of halving steps in the line search optimization.
+        :type max_halving: `int`
+        :param max_doubling: Maximum number of doubling steps in the line search optimization.
+        :type max_doubling: `int`
+        :param batch_size: Internal size of batches on which adversarial samples are generated.
+        :type batch_size: `int`
+        """
+        # Save attack-specific parameters
+        super(CarliniL0Method, self).set_params(**kwargs)
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:1260]:
```suggestion
        super().set_params(**kwargs)
```

File: art/attacks/evasion/carlini.py [hunk 13]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
...
+                        logger.debug("Average L2Dist: %f", np.mean(l2dist))
+                        logger.debug("Average Margin Loss: %f", np.mean(loss - l2dist))
+                        logger.debug(
+                            "Current number of succeeded attacks: %i out of %i",
+                            int(np.sum(attack_success)),
+                            len(attack_success),
+                        )
+
+                        l0dist = np.sum((np.abs(x_batch - x_adv_batch) > self._perturbation_threshold).astype(int),
+                                        axis=1)
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:1022]:
```suggestion
                        l0dist = np.sum((
                            np.abs(x_batch - x_adv_batch) > self._perturbation_threshold).astype(int), axis=1
                        )
```

File: art/attacks/evasion/carlini.py [hunk 14]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
...
+
+                            lr_mult = learning_rate[active_and_do_halving]
+                            for _ in range(len(x.shape) - 1):
+                                lr_mult = lr_mult[:, np.newaxis]
+
+                            x_adv1 = x_adv_batch_tanh[active_and_do_halving]
+                            new_x_adv_batch_tanh = \
+                                x_adv1 + lr_mult * \
+                                perturbation_tanh[do_halving] * \
+                                activation_batch[do_halving]
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:1076]:
```suggestion
                            new_x_adv_batch_tanh = (
                                x_adv1 + lr_mult * perturbation_tanh[do_halving] * activation_batch[do_halving]
                            )
```

File: art/attacks/evasion/carlini.py [hunk 15]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
...
+                            active_and_update_adv[active] = update_adv
+                            best_lr_mult = best_lr[active_and_update_adv]
+                            for _ in range(len(x.shape) - 1):
+                                best_lr_mult = best_lr_mult[:, np.newaxis]
+
+                            x_adv4 = x_adv_batch_tanh[active_and_update_adv]
+                            best_lr1 = best_lr_mult * perturbation_tanh[update_adv]
+                            x_adv_batch_tanh[active_and_update_adv] = \
+                                x_adv4 + best_lr1 * \
+                                activation_batch[active_and_update_adv]
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:1145]:
```suggestion
                            x_adv_batch_tanh[active_and_update_adv] = (
                                x_adv4 + best_lr1 * activation_batch[active_and_update_adv]
                            )
```

File: art/attacks/evasion/carlini.py [hunk 16]
```diff
@@ -797,3 +797,481 @@ def _check_params(self) -> None:
...
+            old_activation[improved_adv] = activation.copy()[improved_adv]
+            activation[improved_adv] *= fix_feature[improved_adv]
+            print(
+                "L0 norm before fixing :\n{}\nNumber active features :\n{}\nIndex of fixed feature :\n{}"
+                .format(
+                    np.sum((perturbation_l1_norm > self._perturbation_threshold).astype(int), axis=1),
+                    np.sum(activation, axis=1),
+                    fix_feature_index
+                )
+            )
```
Comment (beat-buesser) [inline art/attacks/evasion/carlini.py:1231]:
```suggestion
            logger.info(
                "L0 norm before fixing :\n{}\nNumber active features :\n{}\nIndex of fixed feature :\n{}".format(
                    np.sum((perturbation_l1_norm > self._perturbation_threshold).astype(int), axis=1),
                    np.sum(activation, axis=1),
                    fix_feature_index,
                )
            )
```

**Comment by beat-buesser** [comment]
Hi @mamerzouk Thank you very much! I'm pulling this pull request to branch `development_carlini_l0` to add a few last integration changes and will merge it from there to `dev_1.7.0`.
